File: 2561. Rearranging Fruits/solution.cpp
class Solution {
public:
    long long minCost(vector<int>& basket1, vector<int>& basket2) {
        unordered_map<int,int> firstBas;
        unordered_map<int,int> secBas;
        unordered_map<int,int> mixBasket;

        for(int &it : basket1){
            firstBas[it]++;
            mixBasket[it]++;
        }

        for(int &it : basket2){
            secBas[it]++;
            mixBasket[it]++;
        }

        for(auto &[key,value] : mixBasket){
            if(value % 2 != 0){
                return -1;
            }
        }

        vector<int> toSwap1;
        vector<int> toSwap2;


        for(auto &[key,value] : mixBasket){
            int diff = abs(firstBas[key] - secBas[key]) / 2;

            if (firstBas[key] > secBas[key]) {
                while (diff--) toSwap1.push_back(key);
            } else if (secBas[key] > firstBas[key]) {
                while (diff--) toSwap2.push_back(key);
            }
        }

        sort(toSwap1.begin(),toSwap1.end());
        sort(toSwap2.rbegin(), toSwap2.rend());

        int n = toSwap1.size();
        int globalMin = INT_MAX;
        for(int &it : basket1) globalMin = min(globalMin,it);
        for(int &it : basket2) globalMin = min(globalMin,it);

        long long result = 0;

        for(int i=0;i<n;i++){
            result += min(min(toSwap1[i],toSwap2[i]),2*globalMin);
        }

        return result;

    }
};

File: 2561. Rearranging Fruits/solution.java
import java.util.*;

class Solution {
    public long minCost(int[] basket1, int[] basket2) {
        Map<Integer, Integer> firstBas = new HashMap<>();
        Map<Integer, Integer> secBas = new HashMap<>();
        Map<Integer, Integer> mixBasket = new HashMap<>();

        for (int it : basket1) {
            firstBas.put(it, firstBas.getOrDefault(it, 0) + 1);
            mixBasket.put(it, mixBasket.getOrDefault(it, 0) + 1);
        }

        for (int it : basket2) {
            secBas.put(it, secBas.getOrDefault(it, 0) + 1);
            mixBasket.put(it, mixBasket.getOrDefault(it, 0) + 1);
        }

        for (Map.Entry<Integer, Integer> entry : mixBasket.entrySet()) {
            if (entry.getValue() % 2 != 0) {
                return -1;
            }
        }

        List<Integer> toSwap1 = new ArrayList<>();
        List<Integer> toSwap2 = new ArrayList<>();

        for (Map.Entry<Integer, Integer> entry : mixBasket.entrySet()) {
            int key = entry.getKey();
            int diff = Math.abs(firstBas.getOrDefault(key, 0) - secBas.getOrDefault(key, 0)) / 2;

            if (firstBas.getOrDefault(key, 0) > secBas.getOrDefault(key, 0)) {
                for (int i = 0; i < diff; i++) toSwap1.add(key);
            } else if (secBas.getOrDefault(key, 0) > firstBas.getOrDefault(key, 0)) {
                for (int i = 0; i < diff; i++) toSwap2.add(key);
            }
        }

        Collections.sort(toSwap1);
        toSwap2.sort(Collections.reverseOrder());

        int n = toSwap1.size();
        int globalMin = Integer.MAX_VALUE;

        for (int it : basket1) globalMin = Math.min(globalMin, it);
        for (int it : basket2) globalMin = Math.min(globalMin, it);

        long result = 0;

        for (int i = 0; i < n; i++) {
            result += Math.min(Math.min(toSwap1.get(i), toSwap2.get(i)), 2 * globalMin);
        }

        return result;
    }
}
